### ISSUE TITLE
Missing IPv6 addresses - only link local displayed

### ISSUE THREAD

**Issue by tlstpierre** [sebhildebrandt/systeminformation#931]
The networkInterfaces function shows IPv6 address information, but only the first IPv6 address, which seems to be link-local (the least useful address).

This function should return all IPv6 addresses as an array, or at the very least, return the first non-link-local address.  The link-local address probably should be in a different field, or perhaps omitted, since you can calculate it from the MAC anyway.

To reproduce:

-Call the function and log the output on a computer with working IPv6
-Compare the output with the network configuration using ifconfig, ipaddr, or another mechanism to view your network config.
-Notice that only the link local (fe80) address is present, not the globally-unique addresses.

For example:
{
  iface: 'en6',
  ifaceName: 'en6',
  default: true,
  ip4: '172.17.16.146',
  ip4subnet: '255.255.255.0',
  ip6: 'fe80::e:379d:7f60:33a3',
  ip6subnet: 'ffff:ffff:ffff:ffff::',
  mac: 'a0:ce:c8:5f:7e:73',
  internal: false,
  virtual: false,
  operstate: 'up',
  type: 'wired',
  duplex: 'full',
  mtu: 1500,
  speed: 1000,
  dhcp: true,
  dnsSuffix: '',
  ieee8021xAuth: '',
  ieee8021xState: '',
  carrierChanges: 0
}

Compared with the below output (prefixes changed for privacy)
ifconfig en6
en6: flags=8863<UP,BROADCAST,SMART,RUNNING,SIMPLEX,MULTICAST> mtu 1500
	options=6467<RXCSUM,TXCSUM,VLAN_MTU,TSO4,TSO6,CHANNEL_IO,PARTIAL_CSUM,ZEROINVERT_CSUM>
	ether a0:ce:c8:00:00:00
	inet6 fe80::e:379d:7f60:33a3%en6 prefixlen 64 secured scopeid 0xa 
	inet 172.17.16.146 netmask 0xffffff00 broadcast 172.17.16.255
	inet6 2001:db8:10:1:cda:cd4:b581:a96b prefixlen 64 autoconf secured 
	inet6 2001:db8:10:1:bd3f:a40f:e57:8d2c prefixlen 64 autoconf temporary 
	inet6 2001:db8:10:1::8893 prefixlen 64 dynamic 
	nat64 prefix 64:ff9b:: prefixlen 96
	nd6 options=201<PERFORMNUD,DAD>
	media: autoselect (1000baseT <full-duplex>)
	status: active

Tested on Mac OS 12.7.6

Expected behaviour is that the function can return all IPv6 addresses, especially the non-link-local addresses.  The 2001:db8: addresses (normally would be 2xxx:xxxx, but 2001:db8 is the IPv6 example prefix) are the ones the computer uses to communicate outside of the network, so knowing this is typically what one would want to find.

**Comment by sebhildebrandt** [comment]
@tlstpierre ... so the `fe80::` addresses are the link-local addresses, right? As changing from a number value for IP6 to an array would be a breaking change, this is nothing I could do in version 5 ... so my proposal:

- for version 5 of the lib: return back the first non-link-local address
- for version 6 maybe change to returning an array ...

Makes sense for you?

**Comment by tlstpierre** [comment]
Yes, it is typical for interfaces to have multiple addresses with IPv6.  fe80:: is link local, and then you would typically get an EUI-164 address if this is enabled on the network, possibly a dhcpv6 address, and could be even more if privacy extensions are turned on.

This can also be true with IPv4, but is far less common.  If you can move to an array in v6, that is probably the right way to handle it.  In the mean time, if you can only populate a single field, possibly you could implement some simple logic that would skip the link-local address and just display the first globally routable IPv6 address, unless there isn’t one, then show link local.  That would at least make it usable in the sense that you would see an IP address that can be reached.


Tim St. Pierre
***@***.***
289-225-0010

**Comment by sebhildebrandt** [comment]
@tlstpierre I am just thinking ... for V6 would it make sense to create an array with ip-address + type (link.local, ULA, secured, temporary, dynamic)?

**Comment by sebhildebrandt** [comment]
@tlstpierre for V5 I would do it in this hierarchy

- secured 
- dynamic
- temporary
- link-local

The first type then, that I can find will be the one returned as IP6 address. Ok?

**Comment by tlstpierre** [comment]
That sounds pretty sensible.

I also like your suggestion of making the returned values objects that have an IP + type.  A lot of libraries will have methods to determine this from an IP address, so it shouldn’t be too hard to implement.

Thanks!


Tim St. Pierre
***@***.***
289-225-0010

**Comment by sebhildebrandt** [comment]
Hi @tlstpierre ... I just published a new version 5.28.1 which should fix this issue as discussed for version 5. Please feel free to comment your results if you are able to test it on your side.
